### PR TITLE
Add city to build psfs

### PR DESCRIPTION
**Pull request by gonzaponte**
Introduce a city to build psfs. Original code from @Aretno, I just reshaped it into a city.

### CODE REVIEW

**Comment by jacg** [comment]
Erm ... commit 237e32a, the one which adds the liquidata version ... *where* is it? I cannot find it on the `gonzaponte:psfcity` branch, be it via Magit, git CLI or the GitHub web interface.

**Comment by gonzaponte** [comment]
It's in gonzaponte:flatmap. This banch hasn't been rebased yet on top of that. From what I understood [here](https://github.com/next-exp/IC/pull/780#issuecomment-813504031) this is what you wanted.

**Comment by jacg** [comment]
... and it appears *here* because it references this PR. OK. And as it's an implementation of a psf city, the `psfcity` branch seemed the logical place for it to be. Hmm.

File: invisible_cities/config/create_psf_city.conf [hunk 1]
```diff
@@ -0,0 +1,14 @@
+files_in    = '$ICDIR/database/test_data/kr_hits_for_psf.h5'
+file_out    = '/tmp/psf.h5'
+compression = 'ZLIB4'
+
+run_number    = 8000
+detector_db   = 'new'
```
Comment (Aretno) [inline invisible_cities/config/create_psf_city.conf]: I don't see any reason to have that many empty spaces before `=`, dunno if i'm missing something.

File: invisible_cities/cities/create_psf.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                                      , accumulate_psf.sink
+                                      )
+                    , result = accumulate_psf.future
+                    )
+
+    with tb.open_file(file_out, "w") as outfile:
+        write_psf = psf_writer(outfile, compression=compression)
+
+        # TODO: Probably better to ignore this writer
+        # and use df_writer directly
```
Comment (Aretno) [inline invisible_cities/cities/create_psf.py]: Agreed, as you already have the dataframe created in `combine_psfs` you might as well use the dataframe writer.

File: invisible_cities/cities/create_psf.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    return split_in_z
+
+
+def xy_splitter(sectors, bin_edges):
+    def split_in_xy(dst):
+        xsectors, ysectors = sectors
+        for xmin, xmax in zip(xsectors, xsectors[1:]):
+            # TODO: replace floor with something more robust
+            xpsf = np.floor((xmin + xmax) / 2)
+            selx = in_range(dst.X, xmin, xmax)
```
Comment (Aretno) [inline invisible_cities/cities/create_psf.py]: Selection should be made based on Xpeak (and Ypeak a couple lines later)

File: invisible_cities/cities/create_psf.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                     , factor
+                     , entries
+                     )
+
+    return result
+
+
+def column_dropper(columns):
+    def dropper(df):
+        # TODO: is reset_index really needed?
```
Comment (Aretno) [inline invisible_cities/cities/create_psf.py]: I think you are right and it's not needed.

File: invisible_cities/cities/create_psf.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        return df.drop(columns=columns).reset_index(drop=True)
+    return dropper
+
+
+def z_splitter(zbins, ranges, sipms):
```
Comment (Aretno) [inline invisible_cities/cities/create_psf.py]: I would change the name of this function, as it not only splits the data but also does some processing. Actually, `hdst_psf_processing` does not need to necessarily happen post z-splitting so you could have completely separate function doing the processing before doing the splitting. The caveat was that hdst_psf_processing uses groupby and memory comsumption could be delicate. Still as in a city this is executed file by flile it should be ok. I'm more inclined of having separate functions but at the very least go for another name.

File: invisible_cities/cities/create_psf.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    def split_in_z(dst):
+        for zmin, zmax in zip(zbins, zbins[1:]):
+            zpsf = (zmin + zmax) / 2
+
+            sel  = in_range(dst.Z, zmin, zmax)
+            if not np.any(sel): continue
+
+            processed = hdst_psf_processing(dst[sel], ranges, sipms)
+
+            # TODO: is this really needed?
```
Comment (Aretno) [inline invisible_cities/cities/create_psf.py]: I don't think it's needed, nope.

**Review by Aretno — CHANGES_REQUESTED** [review]
Mostly ready to go, aside from some minor fixes (many derived from the original code) I think it would be good to store the run info (or MC info in case of a MC production) in the output file. Additionally, a test checking that two files are properly combined would also be desireable.

**Comment by gonzaponte** [comment]
> I think it would be good to store the run info (or MC info in case of a MC production)

We agreed offline that since this is not done (yet) for any other city and there are no tools to do it, we leave it for another time.

> a test checking that two files are properly combined would also be desireable.

I've modified the exact-result test so it takes a case in which we split in all three coordinates and that runs over three files to make sure we check every possible feature.

I've marked all the other requests as resolved.

**Review by Aretno — COMMENTED** [review]
All good, just name the city (I'm inclined to go with Eutropia) and this is ready to go.

**Review by Aretno — APPROVED** [review]
This PR marks the, long overdue, formalization as a city, Eutropia, of the PSF extraction procedure. It takes penthesilea output (with a big enough rebinning so each sensor charge is fully integrated over time) to produce a set of PSFs, each one corresponding to a different region in the detector. Such regions are configurable. The PSFs are composed of their region position, the relative position sensor-event, the charge fraction seen by the sensor and the number of events used in the computation of such fraction.

Great job adapting the code @gonzaponte.

**Comment by gonzaponte** [comment]
I've just added some type hints and improved alignment here and there. Please merge when we get the green flag from the GHA.

**Pull request merged**